src/Results.Immutable/Member/EachParser.cs

```diff
@@ -0,0 +1,114 @@
+    ///     In case of success of each element, the result will contain the parsed values.
+    /// </summary>
+    /// <param name="values">The values to parse/validate.</param>
+    /// <param name="parser">The parser to use for each element to determine the validity or transform.</param>
+    /// <param name="createError">
+    ///     Creates an any error type based on the element, its index and the errors returned by the
+    ///     parser.
+    /// </param>
+    /// <typeparam name="TElement">The type of the original values.</typeparam>
+    /// <typeparam name="TParsed">The type of the parsed values.</typeparam>
+    /// <returns>A result containing the parsed values or their errors.</returns>
+    public static Result<ImmutableList<TParsed>> ParseEach<TElement, TParsed>(
+        this IEnumerable<TElement> values,
+        Func<TElement, Result<TParsed>> parser,
+        Func<TElement, int, ImmutableList<Error>, Error> createError)
+    {
+        var results = values.Select(
+            (x, i) =>
+            {
+                var result = parser(x);
+                return result.Some is { }
+                    ? result
+                    : Result.Fail<TParsed>(
+                        createError(
+                            x,
+                            i,
+                            result.Errors));
+            });
+
+        return Result.Transpose(results);
+    }
+
+    /// <inheritdoc cref="ParseEach{TElement, TParsed}(IReadOnlyList{TElement}, Func{TElement, Result{TParsed}}, string)" />
+    public static async ValueTask<Result<ImmutableList<TParsed>>> ParseEachAsync<TElement, TParsed>(
+        this IReadOnlyList<TElement> values,
+        Func<TElement, ValueTask<Result<TParsed>>> parser,
+        string message = "") =>
+        await ParseEachAsync(
+            values,
+            parser,
+            (
+                x,
+                i,
+                e) => new IndexError(
+                i,
+                message,
+                e));
+
+    /// <inheritdoc
+    ///     cref="ParseEach{TElement, TParsed}(IEnumerable{TElement}, Func{TElement, Result{TParsed}}, Func{TElement, int, ImmutableList{Error}, Error})" />
+    public static async ValueTask<Result<ImmutableList<TParsed>>> ParseEachAsync<TElement, TParsed>(
+        this IEnumerable<TElement> values,
+        Func<TElement, ValueTask<Result<TParsed>>> parser,
+        Func<TElement, int, ImmutableList<Error>, Error> createError)
+    {
+        var ix = 0;
+        var results = new List<Result<TParsed>>();
+
+        foreach (var item in values)
+        {
+            var result = await parser(item);
+            results.Add(
+                result.Some is { }
+                    ? result
+                    : Result.Fail<TParsed>(
+                        createError(
+                            item,
+                            ix,
+                            result.Errors)));
+
+            ix++;
+        }
```

Comment:
We can abuse quirks of the compiler here (lambda expressions awaiting `ValueTask`s are `Task`s and `ValueTask`s at the same time for the sake of type-checking):

```suggestion
        var results = await Task.WhenAll(
            values.Select(
                async (x, i) =>
                {
                    var result = await parser(x);
                    return result is { Errors: { IsEmpty: false } errors } ?
                        Result.Fail<TParsed>(createError(x, i, errors)) :
                        result;
                }));
```

Reply:
I partially applied your solution, take in mind that using `Error.IsEmpty` may produce an unnecessary allocation if there is no error. It is more efficient to use `.Some`.